File: pybb/model/gaussian.py
import model
import math
import random

class Gaussian(model.Model):
    def __init__(self, mean = 0.0, std = 1.0):
        self.mean = mean
        self.std = std
        self.history = 0
        
    def set(self, mean = 0.0, std = 1.0):
        self.mean = mean
        self.std = std
        
    def forecast(self, data, future = 1):
        return self.mean

    def loglikelihood(self, data):
        """Forecast for a model the probability of each observation.

        f(x) = 1/(std * sqrt(2 * pi)) * exp(-1 * ((x - mean)**2)/(2*std**2))
        """
        nc = 1 / (self.std * math.sqrt(2 * math.pi))
        e = exp(-1 * (((x - self.mean)**2)/(2*(self.std**2))))
        
        return nc * e

File: matlab/arima_forecast/+bcf/+models/Gaussian.m
classdef Gaussian < bcf.models.Model
    %GAUSSIANMODEL Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        mu
        sigma
        history
    end
    
    methods
        function obj = Gaussian(mu, sigma)
            obj.mu = mu;
            obj.sigma = sigma;
            obj.history = 0;
        end
        
        function train(obj, data)
        end
        
        function output = forecastSingle(obj, data, ahead, varargin)
            output = obj.mu;
        end
        
        function output = forecastAll(obj, data, ahead, varargin)
            output = ones(size(data)) .* obj.mu;
        end
            
        function val = probability(obj, data)
            %Forecast for a model the probability of each observation
            %TODO Change this later
            prob = mvnpdf(data, obj.noiseMu, obj.noiseSigma);
        end
        
        
        function calculateNoiseDistribution(obj, data)
            out = obj.forecastAll(data, 1);
            res = data - out;
            obj.noiseSigma = std(res, 1, 2);
            obj.noiseMu = mean(res, 2);
            %pd =  fitdist(res', 'Normal');
            %obj.noiseMu = pd.mean;
            %obj.noiseSigma = pd.std^2;
        end
        
        
        function prob = probabilityNoise(obj, data)
            %Computes the probability that some noise lies within this
            %models forecasted noise

            %First discretize the pdf
            %For now just always go from -2 to 2 by .1
            range = -2:0.01:2;
            dValues = normpdf(range, 0, obj.noiseSigma);
            dValues(dValues < 0.000000001) = 0.000000001;
            dValues = dValues ./ sum(dValues);
            prob = zeros(size(data));
            
            for i = 1:size(data, 2)
                %Change this to include values equal to zero
                foo = max(find(range <= data(1, i))) + 1;
                foo = min([length(dValues), foo]); 
                prob(1, i) = dValues(foo);
            end
            %prob = normpdf(data, obj.noiseMu, obj.noiseSigma);
        end
    end
end

